core/src/main/scala/org/apache/spark/internal/config/package.scala

```diff
@@ -691,6 +691,15 @@ package object config {
     .toSequence
     .createWithDefault(Nil)
 
+  private[spark] val KERBEROS_FILESYSTEM_RENEWAL_EXCLUDE =
+    ConfigBuilder("spark.kerberos.renewal.exclude.hadoopFileSystems")
+      .doc("The list of Hadoop filesystem URLs whose hosts will be excluded from " +
```

Comment:
Is this only restricted to HDFS URLs? can "hosts" either be nameservice name or namenode name? some examples might help.

Reply:
Because the configuration value will be parsed as file system URLs to get file system. A simple host name doesn't work here.

e.g.

```scala
scala> new Path("hdfs.namenode.net").getFileSystem(new Configuration()).getUri.getHost
res9: String = null
```

Reply:
It makes sense for `spark.kerberos.access.hadoopFileSystems` to be URLs since Spark needs to instantiate `FileSystem`s from them. But for this case I'm not sure if it's necessary: we can just parse the config into a set of host names and check whether the file systems above contain them:
```scala
val hostsToExclude = sparkConf.get(KERBEROS_FILESYSTEM_RENEWAL_EXCLUDE).toSet
filesystems.filter(fs => !hostsToExclude.contains(fs.getUri.getHost).foreach { fs =>
 ...
}
```

I'm fine either way though since it also makes sense to keep it consistent with `spark.kerberos.access.hadoopFileSystems`. BTW I think we'll need to update security.md for the new config.

Reply:
Thanks. Let me update security.md.

Reply:
so just to be clear, these fileystems are ones the user specified in spark.kerberos.access.hadoopFileSystems so it wants to get initial tokens for them but then later we don't want to renew, correct?  We should perhaps update doc to mention that.  
It would be nice to know if this works with other cluster managers like k8s - @ifilonenko @mccheah  maybe?

Reply:
I guess it could be the staging or default Filesystem as well.

Reply:
@tgravescs with the actual code Spark doesn't obtain the initial tokens when an FS is listed in "spark.kerberos.renewal.exclude.hadoopFileSystems". Please see my comment below...

Reply:
Hm? @gaborgsomogyi I think this config doesn't affect Spark obtain the initial tokens?

Reply:
Right, the point I was getting at is that we only renew the ones in spark.kerberos.access.hadoopFileSystems + defaultFs + staging fs.  So I assume its basically defaultFs then? I was kind of wanting to make sure it wasn't something silly that they specified in both.  It seems odd for the defaultFs since this is the Hadoop filesystem which supports renewal.  I can obviously see this if its a separate name node but it would require you to specify it in the config for us to try to renew. So I'm assuming this is perhaps an odd config that you can't renew on the current default fs.  Just seems a bit odd case so wanted to make sure I wasn't missing something.

Reply:
Note I'm not against it just making sure the problem is understood and we document it appropriately. I guess it could be stage FS as well, which perhaps might make more sense if on a separate namenoxde

Reply:
It's hard to imagine that this applies to defaultFS/stagingDir since it means the YARN cluster (or other types of clusters) is not configured to be in the same Kerberos realm as these which could cause other more serious issues. But, just as `spark.kerberos.access.hadoopFileSystems`, I guess nothing stops users from putting in the host for defaultFS/stagingDir in the config as well, and Spark will just do what it's told to do.

Reply:
From the user perspective, I think it is still possible they want to avoid renewal for defaultFs. If the defaultFs is just a remote HDFS that whose tokens cannot be renew from YARN?

Reply:
yes its true like I mentioned above. I'm not against it just want to understand and make sure this config makes sense. For instance instead of having this config we could have one that says don't include defaultFs or stageFs or both in renewal.   I think those are really the only 2 filesystems that make sense to put into this config. Anything else doesn't get renewed unless its in spark.kerberos.access.hadoopFileSystems and in that case just don't include it in spark.kerberos.access.hadoopFileSystems.  The nice things about this approach is its generic if things change in the future. The downside is I have to know namenode path for the defaultFs and staging fs to add to this config.

Reply:
I'm not sure I understand correctly so just consider it as ignorant question.

> Anything else doesn't get renewed unless its in spark.kerberos.access.hadoopFileSystems and in that case just don't include it in spark.kerberos.access.hadoopFileSystems.

Is it possible if someone wants to get initial delegation token for them but doesn't renew them, and does this change make difference from here?

Reply:
If I understand correctly, it is the same code path for obtaining delegation token and renewing. Actually the so called renew is that Spark periodically obtain delegation token.

So I guess it doesn't make sense for a case to get delegation token but not renew it. At least in Spark I don't see such case is possible. I may wrong here.

Reply:
@HeartSaVioR Spark actually behaves this way (w/o this change): if one DT is created then it's renewed all the time except in the following cases:
* `keytab` used but no `principal` provided
* `ccache` used but it has no kerberos credentials

Please see the code here: https://github.com/apache/spark/blob/a916690dd9aac40df38922dbea233785354a2f2a/core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala#L83-L88

Reply:
Ah OK I think I missed some part of @tgravescs 's comment. My bad.

His point wasn't that the config is not necessary at all. His point was that the problem will occur from only defaultFs and/or stageFs, so the new config could be simplified instead of being general but a bit verbose.

Reply:
> so just to be clear, these fileystems are ones the user specified in spark.kerberos.access.hadoopFileSystems so it wants to get initial tokens for them but then later we don't want to renew, correct? We should perhaps update doc to mention that.
> It would be nice to know if this works with other cluster managers like k8s - @ifilonenko @mccheah maybe?

Hmm, for the customer case, the requirement is to prevent YARN from renewing the delegation token obtained. It doesn't matter if the file system is defaultFs/stageFs or the ones specified in `spark.kerberos.access.hadoopFileSystems`. That said, Spark still can obtain the token for the ones in `spark.kerberos.access.hadoopFileSystems`, but we don't want YARN to renew it.

Reply:
I was simply trying to clarify the exact case being hit here and make sure there wasn't an alternate solution and make sure the docs are clear.   The specific case being hit could affect the solution.

I think there are comments on this review that are very confusing, thus why I wanted to clarification. Some indicate that Spark doesn't get initial tokens, others saying in this case the tokens were already acquired, etc.

In the end my comment ends up being I think we should update the security.md doc to mention renewal in the kerberos section for Hadoop filesystems to help explain to users.

Reply:
Okay, thanks @tgravescs. I think in short this config doesn't affect Spark token handling, but only prevent YARN from renewing tokens, no matter the tokens are obtained in advance or by Spark. Let me update security.md and try to explain it.